man/man1/catclm.1
.TH catclm 1  "USER COMMANDS"
.SH NAME
catclm \- concatenate LPJmL climate data files
.SH SYNOPSIS
.B catclm
[\-v] [\-f] [\-longheader] [\-metafile] [\-json] [\-size4]
\fIinfile\fP ... \fIoutfile\fP
.SH DESCRIPTION
Program concatenates climate data files in clm format.
.SH OPTIONS
.TP
\-v
Print filename and start end year of each input file.
.TP
\-f
Force overwrite of existing output file.
.TP
\-longheader
Long (version 2) header assumed for clm files.
.TP
\-metafile
Input files are JSON metafiles describing the structure of the data. Option implicitely sets \-json option and an additional JSON metafile with suffix \fI.json\fP is written for output file.
.TP
\-json
Additional JSON metafile is written.
.TP
\-size4
Size of data is set to 4 bytes for version 2 clm files, default is 2 bytes.
.TP
.I infile
Filename(s) of clm files to concatenate.
.TP
.I outfile
Filename of concatenated clm file.
.SH EXAMPLE
.TP
Create clm data file from two clm files:
.B catclm
temp_1901-2003.clm temp_2004-2055.clm temp_1901-2055.clm
.PP
.SH EXIT STATUS
.B catclm
returns a zero exit status if the clm file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
lpjml(1), mathclm(1), cutclm(1), txt2clm(1), mergeclm(1), grid2clm(1), cft2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security


## [Unreleased]


## [6.1.4] - 2026-07-03

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Fabian Stenzel (stenzel@pik-potsdam.de), David Hötten (davidho@pik-potsdam.de)

### Added

- Utility `cmprestart` added to compare two restart files and check for identity.
- Function `bstruct_skiparray()` and corresponding test `test_skiparray.c` added in `bstruct` library to skip arrary elements.
- Function `bstruct_gettype()` added to `bstruct` library to get datatype of object in restart file.
- Function `bstrucht_printnamestack()` added for more comprehensive error messages.
- Comparison of restart files with `cmprestart` added to merge request template.

### Changed

- Index array object is always skipped in `bstruct_findobject.c`, test `test_indexarray.c` updated to test for reading array without seeking to an index array.
- Check for seeking to position within file added to `bstruct_findobject.c`, `bstruct_open.c`, `bstruct_wopen.c` and `bstruct_seekindexarray.c`.
- Constant `MAXLEVEL` renamed to `BSTRUCT_MAXLEVEL`.
- Macros `getname()` and `isinvalidtoken()` renamed to `bstruct_getname()` and `bstruct_isinvalidtoken()`.
- String comparison in `bstruct_fprintdata` replaced by call to `isnan()` and `isinf()`.
- New token `bstruct_dzero` defined for restart files to distinguish between float and double zeros, version of bstruct files changed to 2.


## [6.1.3] - 2026-07-03

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth(wirth@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Jannes Breier (breier@pik-potsdam.de)

### Added

- Option `-metafile` and `-json` added to `cutclm` and `catclm`.
- Option `-double` added to `cdf2soil.c` to write grid file with double datatype.

### Changed

- New datatype `Metadata` defined in `types.h` combining maps, attributes, basetemp and hlimit arrays used by `opendata()`.
- basetemp, countrymap and hlimit arrays are copied from JSON metafiles to newly created metafiles in `regridclm`, `cutlcm`, `catclm`, and `mathclm`.
- Scaling to sum of stand fractions removed for output `"vegc_avg"`.
- Utility `getcountry` reads now JSON metafile for country codes including the alpha-3 country codes and global grid instead of CLM file.
- If `-pedantic` option is set, non-matching titles in input files stops `lpjml` with an error.
- Datatype of `lon`, `lat`, and `area` elements of `Coord` changed from `Real` to `double`.
- Missing path to grid file added in `openmetafile.c`.

### Fixed

- Warning for float coordinates corrected in `cdf2coord.c`, `cdf2grid.c` and `cdf2soil.c`.
- Variables initialized to zero in `reservoir2cdf.c` and `restart2yaml.c` to avoid warnings by `gcc`.
- Carbon balance fixed in `reclaim_land.c`.
- Check for NULL pointer added in `getuser.c` to avoid SEGV if user name cannot be resolved.
- Verbose output corrected for input without time axis in `cdf2clm.c`.
- Output is rescaled by the actual sum of stand fractions in `update_daily_cell.c` and `fwriteoutput.c` to avoid division by zero (issue #411).
- Default timestep set to 1 in `regridclm.c`.
- Variable `depth` initialized to zero before layer loop in `fwriteoutput.c` to correct `soilc_1m` output.
- Length of buffer for hostname corrected in `gethost.c`.
- Output of unprintable characters corrected in `fputprintable.c`.
- Memory leak fixed in `openmetafile.c` in case of error.


## [6.1.2] - 2026-06-16

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Fixed

- Check for last year for lightning and human ignition disabled in `filesexist.c`, `lpjcheck` does not return an error for standard input.


## [6.1.1] - 2026-06-12

### Contributors

author: Jannes Breier (breier@pik-potsdam.de),  Werner von Bloh (bloh@pik-potsdam.de)
code review: Christoph Müller (cmueller@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)

### Added

- `logo.png` added to `man` directory
- `LICENSES` folder
- `LICENSES/CC0-1.0.md` added for artwork license

### Changed

- `AUTHORS` renamed to `AUTHORS.md` and restructured for active and previous authors
- `INSTALL` renamed to `INSTALL.md` and adjusted
- `LICENSE` renamed to `LICENSES/AGPL-3.0.md` and moved to `LICENSES` folder
- `README` renamed to `README.md` and expanded
- `configure.sh`, `src/lpj/copyright.c` and `src/lpj/printlicense.c` adjusted to reflect renaming of files
- Updated module version for `oneAPI` in `run_simulate_default.sh`


## [6.1.0] - 2026-06-12

### Contributors

- author: Luke Oberhagemann (lukeober@pik-potsdam.de), Markus Drüke (Markus.Drueke@dwd.de), Maik Billing (billing@pik-potsdam.de), Werner von Bloh (bloh@pik-potsdam.de)
- code review: Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Added

- Option `-inpath` added to `configure.sh` to set path for input files.
- Function `fprintclimbuf()` added in `climbuf.c` to print variables and is called in `fprintcell.c`.
- Function `fprintbuffer()` added in `buffer.c` to print variables and is called in `fprintclimbuf.c`.
- Lakes setting stored in restart file and setting compared to actual setting in `openrestart.c`.

src/utils/catclm.c
/**************************************************************************************/
/**                                                                                \n**/
/**                      c  a  t  c  l  m  .  c                                    \n**/
/**                                                                                \n**/
/**     Concatenates LPJ climate data files                                        \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"
#include <sys/stat.h>
#define USAGE "Usage: catclm [-f] [-v] [-longheader] [-metafile] [-json] [-size4] infile1.clm [infile2.clm ...] outfile.clm\n"

int main(int argc,char **argv)
{
  Header header,oldheader={};
  Metadata metadata,metadata2;
  String id;
  Type grid_type;
  Filename grid_name;
  int i,j,k,firstyear=0,version,n,setversion,iarg,firstversion=0,format;
  FILE *in,*out;
  short *values;
  char *arglist,*out_json;
  int *ivals;
  int *index=NULL,*index2=NULL;
  Byte *bvals;
  long long *lvals;
  struct stat filestat;
  Bool swap,verbose,force,ismeta=FALSE,isjson=FALSE,rc;
  size_t size,filesize,offset;
  char c;
  size=2;
  setversion=READ_VERSION;
  verbose=FALSE;
  force=FALSE;
  grid_name.name=NULL;
  format=CLM;
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-longheader"))
        setversion=2;
      else if(!strcmp(argv[iarg],"-f"))
        force=TRUE;
      else if(!strcmp(argv[iarg],"-v"))
        verbose=TRUE;
      else if(!strcmp(argv[iarg],"-metafile"))
        ismeta=TRUE;
      else if(!strcmp(argv[iarg],"-json"))
        isjson=TRUE;
      else if(!strcmp(argv[iarg],"-size4"))
        size=4;
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n"
                USAGE,argv[iarg]);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(argc<iarg+2)
  {
    fprintf(stderr,"Error: Argument(s) missing.\n"
                   USAGE);
    return EXIT_FAILURE;
  }
  n=argc-iarg-1;
  if(!force)
  {
    if(!stat(argv[argc-1],&filestat))
    {
      fprintf(stderr,"File '%s' already exists, overwrite (y/n)?\n",argv[argc-1]);
      scanf("%c",&c);
      if(c!='y')
        return EXIT_FAILURE;
    }
  }
  out=fopen(argv[argc-1],"wb");
  if(out==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s\n",argv[argc-1],strerror(errno));
    return EXIT_FAILURE;
  }
  initmetadata(&metadata,NULL);
  initmetadata(&metadata2,NULL);
  for(i=0;i<n;i++)
  {
    if(ismeta)
    {
      /* set default values */
      header.datatype=LPJ_SHORT;
      header.timestep=1;
      header.nbands=1;
      header.nstep=1;
      header.order=CELLYEAR;
      header.firstcell=0;
      header.firstyear=1901;
      header.cellsize_lon=header.cellsize_lat=0.5;
      header.ncell=1;
      header.nyear=1;
      version=4;
      grid_type=LPJ_SHORT;
      if(i==0)
        in=openmetafile(&header,&metadata,&grid_name,&grid_type,&format,&swap,&offset,argv[iarg+i],TRUE);
      else
      {
        initmetadata(&metadata2,NULL);
        in=openmetafile(&header,&metadata2,NULL,&grid_type,NULL,&swap,&offset,argv[iarg+i],TRUE);
      }
      if(in==NULL)
      {
        free(grid_name.name);
        freemetadata(&metadata);
        fclose(out);
        return EXIT_FAILURE;
      }
      if(i==0 && format==CLM)
      {
        if(freadheaderid(in,id,TRUE))
        {
          free(grid_name.name);
          freemetadata(&metadata);
          freemetadata(&metadata2);
          fclose(in);
          fclose(out);
          return EXIT_FAILURE;
        }
      }
      fseek(in,offset,SEEK_SET);
      size=typesizes[header.datatype];
    }
    else
    {
      in=fopen(argv[iarg+i],"rb");
      if(in==NULL)
      {
        fprintf(stderr,"Error opening '%s': %s\n",argv[iarg+i],strerror(errno));
        fclose(out);
        return EXIT_FAILURE;
      }
      version=setversion;
      if(i)
        rc=freadheader(in,&header,&swap,id,&version,TRUE);
      else
        rc=freadanyheader(in,&header,&swap,id,&version,TRUE);
      if(rc)
      {
        fprintf(stderr,"Error reading header in '%s'.\n",argv[i+iarg]);
        fclose(in);
        fclose(out);
        return EXIT_FAILURE;
      }
      if(version>CLM_MAX_VERSION)
      {
        fprintf(stderr,"Error: Unsupported version %d in '%s', must be less than %d.\n",
                version,argv[i+iarg],CLM_MAX_VERSION+1);
        fclose(in);
        fclose(out);
        return EXIT_FAILURE;
      }
      filesize=getfilesizep(in)-headersize(id,version);
      if((header.order==CELLINDEX && filesize!=sizeof(int)*header.ncell+((version>=3) ? typesizes[header.datatype] : size)*header.ncell*header.nbands*header.nstep*header.nyear) ||
         (header.order!=CELLINDEX && filesize!=((version>=3) ? typesizes[header.datatype] : size)*header.ncell*header.nbands*header.nstep*header.nyear))
      {
        fprintf(stderr,"Error: file length of '%s' does not match header.\n",argv[iarg+i]);
        fclose(in);
        fclose(out);
        return EXIT_FAILURE;
      }
    }
    if(header.order==CELLINDEX)
    {
      if(i==0)
      {
        index=newvec(int,header.ncell);
        if(index==NULL)
        {
          printallocerr("index");
          free(grid_name.name);
          freemetadata(&metadata);
          freemetadata(&metadata2);
          fclose(in);
          fclose(out);
          return EXIT_FAILURE;
        }
      }
      if(freadint(index,header.ncell,swap,in)!=header.ncell)
      {
        fprintf(stderr,"Error reading cell index in '%s'.\n",argv[i+iarg]);
          free(grid_name.name);
        freemetadata(&metadata);
        freemetadata(&metadata2);
        fclose(in);
        fclose(out);
        return EXIT_FAILURE;
      }
    }
    if(i)
    {
      if(version>=3 && header.datatype!=oldheader.datatype)
      {
        fprintf(stderr,"Error: Different datatype in '%s'.\n",argv[i+iarg]);
        free(grid_name.name);
        freemetadata(&metadata);
        freemetadata(&metadata2);
        fclose(in);
        fclose(out);
        return EXIT_FAILURE;
      }
      if(header.order!=oldheader.order)
      {
        fprintf(stderr,"Error: Different order in '%s'.\n",argv[i+iarg]);
        free(grid_name.name);
        freemetadata(&metadata);
        freemetadata(&metadata2);
        fclose(in);
        fclose(out);
        return EXIT_FAILURE;
      }
      if(header.ncell!=oldheader.ncell)
      {
        fprintf(stderr,"Error: Different number of cells in '%s'.\n",argv[i+iarg]);
        free(grid_name.name);
        freemetadata(&metadata);
        freemetadata(&metadata2);
        fclose(in);
        fclose(out);
        return EXIT_FAILURE;
      }
      if(header.firstcell!=oldheader.firstcell)
      {
        fprintf(stderr,"Error: Different index of first cell in '%s'.\n",argv[i+iarg]);
        free(grid_name.name);
        freemetadata(&metadata);
        freemetadata(&metadata2);
        fclose(in);
        fclose(out);
        return EXIT_FAILURE;
      }
      if(header.nbands!=oldheader.nbands)
      {
        fprintf(stderr,"Error: Different number of bands in '%s'.\n",argv[i+iarg]);
        free(grid_name.name);
        freemetadata(&metadata);
        freemetadata(&metadata2);
        fclose(in);
        fclose(out);
        return EXIT_FAILURE;
      }
      if(header.nstep!=oldheader.nstep)
      {
        fprintf(stderr,"Error: Different number of steps in '%s'.\n",argv[i+iarg]);
        free(grid_name.name);
        freemetadata(&metadata);
        freemetadata(&metadata2);
        fclose(in);
        fclose(out);
        return EXIT_FAILURE;
      }
      if(header.firstyear!=oldheader.firstyear+oldheader.nyear)
      {
        fprintf(stderr,"Error: First year=%d in '%s' is different from %d.\n",header.firstyear,argv[i+iarg],oldheader.firstyear+oldheader.nyear);
        free(grid_name.name);
        freemetadata(&metadata);
        freemetadata(&metadata2);
        fclose(in);
        fclose(out);
        return EXIT_FAILURE;
      }
      if(header.order==CELLINDEX)
      {
        for(j=0;j<header.ncell;j++)
          if(index[j]!=index2[j])
          {
            fprintf(stderr,"Cell index different in '%s'.\n",argv[i+iarg]);
            free(grid_name.name);
            freemetadata(&metadata);
            freemetadata(&metadata2);
            fclose(in);
            fclose(out);
            return EXIT_FAILURE;
          }
      }
      if(ismeta)
      {
        if(metadata.unit!=NULL && metadata2.unit!=NULL && strcmp(metadata.unit,metadata2.unit))
          fprintf(stderr,"Warning: Unit '%s' in '%s' differs from '%s' in '%s'.\n",
                  metadata2.unit,argv[iarg+i],metadata.unit,argv[iarg]);
        freemetadata(&metadata2);
      }
    }
    else
    {
      fseek(out,headersize(id,version),SEEK_SET);
      if(header.order==CELLINDEX)
      {
        index=newvec(int,header.ncell);
        if(index==NULL)
        {
          printallocerr("index");
          free(grid_name.name);
          freemetadata(&metadata);
          fclose(in);
          fclose(out);
          return EXIT_FAILURE;
        }
        if(freadint(index,header.ncell,swap,in)!=header.ncell)
        {
          fprintf(stderr,"Error reading cell index in '%s'.\n",argv[i+iarg]);
          free(grid_name.name);
          freemetadata(&metadata);
          fclose(in);
          fclose(out);
          return EXIT_FAILURE;
        }
        if(fwrite(index,sizeof(int),header.ncell,out)!=header.ncell)
        {
          fprintf(stderr,"Error writing to '%s'.\n",argv[argc-1]);
          free(grid_name.name);
          freemetadata(&metadata);
          fclose(in);
          fclose(out);
          return EXIT_FAILURE;
        }
        index2=newvec(int,header.ncell);
        if(index2==NULL)
        {
          printallocerr("index");
          free(grid_name.name);
          freemetadata(&metadata);
          fclose(in);
          fclose(out);
          return EXIT_FAILURE;
        }
        for(j=0;j<header.ncell;j++)
          index2[j]=index[j];
      }
      firstyear=header.firstyear;
      firstversion=version;
    }
    if(verbose)
      printf("Filename: %s, %d-%d\n",argv[i+iarg],header.firstyear,header.firstyear+header.nyear-1);
    if(version>=3)
      size=typesizes[header.datatype];
    switch(size)
    {
      case 1:
        bvals=newvec(Byte,(long long)header.nbands*header.nstep*header.ncell);
        check(bvals);
        for(j=0;j<header.nyear;j++)
        {
          if(fread(bvals,1,(long long)header.nbands*header.nstep*header.ncell,in)!=(long long)header.nbands*header.nstep*header.ncell)
          {
            fprintf(stderr,"Error reading from '%s'.\n",argv[i+iarg]);
            free(grid_name.name);
            freemetadata(&metadata);
            fclose(in);
            fclose(out);
            return EXIT_FAILURE;
          }
          if(fwrite(bvals,1,(long long)header.nbands*header.nstep*header.ncell,out)!=(long long)header.nbands*header.nstep*header.ncell)
          {
            fprintf(stderr,"Error writing to '%s'.\n",argv[argc-1]);
            free(grid_name.name);
            freemetadata(&metadata);
            fclose(in);
            fclose(out);
            return EXIT_FAILURE;
          }
        }
        free(bvals);
        break;
      case 2:
        values=newvec(short,(long long)header.nbands*header.nstep*header.ncell);
        check(values);
        for(j=0;j<header.nyear;j++)
        {
          if(fread(values,sizeof(short),(long long)header.nbands*header.nstep*header.ncell,in)!=(long long)header.nbands*header.nstep*header.ncell)
          {
            fprintf(stderr,"Error reading from '%s'.\n",argv[i+iarg]);
            free(grid_name.name);
            freemetadata(&metadata);
            fclose(in);
            fclose(out);
            return EXIT_FAILURE;
          }
          if(swap)
            for(k=0;k<(long long)header.nbands*header.nstep*(long long)header.ncell;k++)
              values[k]=swapshort(values[k]);
          if(fwrite(values,sizeof(short),(long long)header.nbands*header.nstep*header.ncell,out)!=(long long)header.nbands*header.nstep*header.ncell)
          {
            fprintf(stderr,"Error writing to '%s'.\n",argv[argc-1]);
            free(grid_name.name);
            freemetadata(&metadata);
            fclose(in);
            fclose(out);
            return EXIT_FAILURE;
          }
        }
        free(values);
        break;
      case 4:
        ivals=newvec(int,(long long)header.nbands*header.nstep*header.ncell);
        check(ivals);
        for(j=0;j<header.nyear;j++)
        {
          if(fread(ivals,sizeof(int),(long long)header.nbands*header.nstep*header.ncell,in)!=(long long)header.nbands*header.nstep*header.ncell)
          {
            fprintf(stderr,"Error reading from '%s'.\n",argv[i+iarg]);
            free(grid_name.name);
            freemetadata(&metadata);
            fclose(in);
            fclose(out);
            return EXIT_FAILURE;
          }
          if(swap)
            for(k=0;k<(long long)header.nbands*header.nstep*header.ncell;k++)
              ivals[k]=swapint(ivals[k]);
          if(fwrite(ivals,sizeof(int),(long long)header.nbands*header.nstep*header.ncell,out)!=(long long)header.nbands*header.nstep*header.ncell)
          {
            fprintf(stderr,"Error writing to '%s'.\n",argv[argc-1]);
            free(grid_name.name);
            freemetadata(&metadata);
            fclose(in);
            fclose(out);
            return EXIT_FAILURE;
          }
        }
        free(ivals);
        break;
      case 8:
        lvals=newvec(long long,(long long)header.nbands*header.nstep*header.ncell);
        check(lvals);
        for(j=0;j<header.nyear;j++)
        {
          if(fread(lvals,sizeof(long long),(long long)header.nbands*header.nstep*header.ncell,in)!=(long long)header.nbands*header.nstep*header.ncell)
          {
            fprintf(stderr,"Error reading from '%s'.\n",argv[i+iarg]);
            free(grid_name.name);
            freemetadata(&metadata);
            fclose(in);
            fclose(out);
            return EXIT_FAILURE;
          }
          if(swap)
            for(k=0;k<(long long)header.nbands*header.nstep*header.ncell;k++)
              lvals[k]=swaplong(lvals[k]);
          if(fwrite(lvals,sizeof(long long),(long long)header.nbands*header.ncell,out)!=(long long)header.nbands*header.nstep*header.ncell)
          {
            fprintf(stderr,"Error writing to '%s'.\n",argv[argc-1]);
            free(grid_name.name);
            freemetadata(&metadata);
            fclose(in);
            fclose(out);
            return EXIT_FAILURE;
          }
        }
        free(lvals);
        break;

    } /* of 'switch' */
    fclose(in);
    oldheader=header;
  }
  free(index);
  free(index2);
  header.nyear=header.firstyear+header.nyear-firstyear;
  header.firstyear=firstyear;
  rewind(out);
  fwriteheader(out,&header,id,firstversion);
  fclose(out);
  if(ismeta || isjson)
  {
    out_json=malloc(strlen(argv[argc-1])+strlen(JSON_SUFFIX)+1);
    if(out_json==NULL)
    {
      printallocerr("filename");
      freemetadata(&metadata);
      free(grid_name.name);
      return EXIT_FAILURE;
    }
    strcat(strcpy(out_json,argv[argc-1]),JSON_SUFFIX);
    arglist=catstrvec(argv,argc);
    if(arglist==NULL)
    {
      printallocerr("arglist");
      free(out_json);
      freemetadata(&metadata);
      free(grid_name.name);
      return EXIT_FAILURE;
    }
    out=fopen(out_json,"w");
    if(out==NULL)
    {
      printfcreateerr(out_json);
      free(out_json);
      free(arglist);
      freemetadata(&metadata);
      free(grid_name.name);
      return EXIT_FAILURE;
    }
    fprintjson(out,argv[argc-1],NULL,arglist,&header,&metadata,(grid_name.name==NULL) ? NULL : &grid_name,grid_type,format,id,FALSE,max(version,(ismeta) ? 4 : 3));
    free(out_json);
    free(arglist);
    fclose(out);
  }
  freemetadata(&metadata);
  free(grid_name.name);
  return EXIT_SUCCESS;
} /* of 'main' */
